### PR TITLE
Add an editor tool to automatically upgrade and re-save meshes

### LINKED ISSUE #83287: Warning about upgrading mesh format to 4.2 is unclear about next steps, and there's no visual feedback when the conversion takes a long time
### Godot version

v4.2.beta.custom_build [b1371806a]

### System information

Godot v4.2.beta (b1371806a) - Windows 10.0.19045 - Vulkan (Forward+) - dedicated NVIDIA GeForce RTX 2060 SUPER (NVIDIA; 31.0.15.3758) - AMD FX(tm)-8350 Eight-Core Processor (8 Threads)

### Issue description

I'm opening a large 3d project and Godot was unresponsive for several minutes.

The console showed the following:
image

I do get the warnings in the debugger but only when the editor unfreezes itself.

I understand this could be considered an enhancement but to me it seems like a UX bug. Launching the editor without a console has been possible for a while and it's the default on platforms like Steam. For non-console users the experience is an unresponsive window with no idea of what's happening. If this were intended behavior there would be no output on the console either. 

### Steps to reproduce

- Launch a project created in a version before 4.2 including a large amount of meshes

Expected: I get some sort of progress bar in the editor while the meshes are converted to the new surface format, as well as the warning about backwards compatibility

Actual: the editor hangs for minutes. No visual indication of what it's doing

### Minimal reproduction project

N/A

### PR DESCRIPTION
**Pull request by clayjohn**
Fixes: https://github.com/godotengine/godot/issues/83287

This PR adds a popup the first time a surface that needs to be upgraded is encountered. The user can select "Upgrade" to have the engine automatically re-import/re-save all meshes so they are saved with the new format. 

Screenshot from 2023-10-19 16-09-05

Screenshot from 2023-10-19 16-09-15

Screenshot from 2023-10-19 16-10-14

Also, changes the warning message (when p_path is available) to:
```
A surface of " + p_path + " uses an old surface format and needs to be upgraded. The upgrade happens automatically at load time every time until the mesh is saved again or re-imported. Once saved (or re-imported), this mesh will be incompatible with earlier versions of Godot.
```

### CODE REVIEW

File: editor/surface_upgrade_dialog.h [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+/**************************************************************************/
+
+#ifndef SURFACE_UPGRADE_DIALOG_H
+#define SURFACE_UPGRADE_DIALOG_H
+
+#include "scene/main/node.h"
+
+class EditorFileSystemDirectory;
+
+class SurfaceUpgradeDialog : public Object {
```
Comment (YuriSizov) [inline editor/surface_upgrade_dialog.h:38]: It should probably not be called a dialog if it's not a window or a control node. `SurfaceUpgradeManager`? Or something along those lines.

Comment (YuriSizov) [reply editor/surface_upgrade_dialog.h:38]: Does it need to be a Godot object by the way?

File: editor/surface_upgrade_dialog.cpp [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+		if (res.is_valid()) {
+			// Ignore things that fail to load.
+			ResourceSaver::save(res);
+		}
+	}
+}
+
+void SurfaceUpgradeDialog::_show_popup() {
+	RS::get_singleton()->set_surface_upgrade_callback(nullptr);
+	if (EditorNode::immediate_confirmation_dialog(TTR("Attempting to use a mesh with an outdated mesh format. \nThe engine needs to update the format in order to use the mesh.\n\nPress 'Upgrade & Re-save' to have the engine scan the project folder and\nautomatically update and re-save all meshes and scenes.\nThis update may take a few minutes.\nUpgrading will make the meshes incompatible with previous versions of Godot.\n\nPress 'Upgrade Only' to continue opening the scene as normal. The engine will update\neach mesh in memory, but the update will not be saved.\nChoosing this option will lead to slower load times every time this project is loaded."), TTR("Upgrade & Re-save"), TTR("Upgrade Only"))) {
```
Comment (YuriSizov) [inline editor/surface_upgrade_dialog.cpp:81]: Maybe move it to a variable so the conditional is more readable? 🙃

File: editor/surface_upgrade_dialog.cpp [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+			}
+		}
+	}
+}
+
+void SurfaceUpgradeDialog::upgrade_all_meshes() {
+	// Update all meshes here.
+	HashSet<String> paths;
+
+	PackedStringArray files_to_import = PackedStringArray();
```
Comment (YuriSizov) [inline editor/surface_upgrade_dialog.cpp:59]:
```suggestion
	PackedStringArray files_to_import;
```

File: editor/surface_upgrade_dialog.cpp [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+
+	_add_files(EditorFileSystem::get_singleton()->get_filesystem(), paths, files_to_import);
```
Comment (YuriSizov) [inline editor/surface_upgrade_dialog.cpp:61]: Can probably remove empty lines in between since this is all related.

File: editor/surface_upgrade_dialog.cpp [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#include "surface_upgrade_dialog.h"
+
+#include "editor/editor_file_system.h"
+#include "editor/editor_node.h"
+#include "servers/rendering_server.h"
+
+void SurfaceUpgradeDialog::_add_files(EditorFileSystemDirectory *p_dir, HashSet<String> &p_paths, PackedStringArray &p_files) {
```
Comment (YuriSizov) [inline editor/surface_upgrade_dialog.cpp:37]: Should probably use the `r_` prefix for the last two arguments.

**Review by YuriSizov — APPROVED** [review]
I haven't tested it, but from the code point of few it looks good, aside from a couple nits and small suggestions.

File: editor/surface_upgrade_dialog.h (hunk 1 above)
Comment (clayjohn) [reply editor/surface_upgrade_dialog.h:38]: Heh, I named it SurfaceUpgradeManager first, then Dialog, then Tool and finally back to Dialog. I don't have a strong preference between any of them.

**Comment by noidexe** [comment]
I tested importing a project with 4.1.1 and then opening it with the windows-editor artifact generated by this PR and it seems to work well overall.

I only had an issue with a large gltf file, getting a "Safe save failed" error message. I could reproduce it on Windows 10 and Windows 11, even with Windows Defender disabled. Disabling safe save and reloading the project to re-trigger the conversion worked well. Even with safe save active, just clicking on reimport after the error also seemed to work well. 

The rest of the gltf files converted flawlessly so it's probably an unrelated issue. In terms of UX I think everything's  much clearer now.

File: editor/surface_upgrade_dialog.cpp (hunk 2 above)
Comment (akien-mga) [reply editor/surface_upgrade_dialog.cpp:81]:
```suggestion
	if (EditorNode::immediate_confirmation_dialog(TTR("Attempting to use a mesh with an outdated mesh format.\nThe engine needs to update the format in order to use the mesh.\n\nPress 'Upgrade & Re-save' to have the engine scan the project folder and\nautomatically update and re-save all meshes and scenes.\nThis update may take a few minutes.\nUpgrading will make the meshes incompatible with previous versions of Godot.\n\nPress 'Upgrade Only' to continue opening the scene as normal. The engine will update\neach mesh in memory, but the update will not be saved.\nChoosing this option will lead to slower load times every time this project is loaded."), TTR("Upgrade & Re-save"), TTR("Upgrade Only"))) {
```
Trailing space.

But the manual line breaks are also problematic and won't necessarily be translated properly, or scale well with different screen resolutions or editor scaling.

It would be better to use autowrapping, and limit the width of the dialog.

**Comment by lostminds** [comment]
As reported here https://github.com/godotengine/godot/issues/83429 the importer on 4.2 will change the names of some meshes (due to previous bug with mesh name that is now fixed) breaking references. Could this tool be a good place to detect/warn or even deal with this as well when the reimport happens?

File: editor/surface_upgrade_dialog.cpp (hunk 2 above)
Comment (YuriSizov) [reply editor/surface_upgrade_dialog.cpp:81]: It appears to be wrapped already, so all that is needed is to remove single newlines within paragraphs. Splitting paragraphs with doubled newlines is fine, and we do it in the project converter dialogs.

**Comment by YuriSizov** [comment]
@lostminds This is tightly related to checks done in the rendering server, which has nothing to do with the importer. So without a completely different implementation this is not compatible with your suggestion.

File: scene/resources/mesh.cpp [hunk 6]
```diff
@@ -1638,7 +1638,7 @@ void ArrayMesh::_set_surfaces(const Array &p_surfaces) {
 #ifndef DISABLE_DEPRECATED
 		uint64_t surface_version = surface.format & (ARRAY_FLAG_FORMAT_VERSION_MASK << ARRAY_FLAG_FORMAT_VERSION_SHIFT);
 		if (surface_version != ARRAY_FLAG_FORMAT_CURRENT_VERSION) {
-			RS::_fix_surface_compatibility(surface);
+			RS::get_singleton()->_fix_surface_compatibility(surface, get_path());
```
[the remark this answers is not in the document]
Comment (clayjohn) [reply scene/resources/mesh.cpp:1641]: I guess so. We use underscore for internal helper methods, which this started out as. But since its being used outside the RenderingServer now, it makes sense to rename the function

File: editor/surface_upgrade_dialog.cpp (hunk 2 above)
Comment (clayjohn) [reply editor/surface_upgrade_dialog.cpp]: Done!

**Review by akien-mga — APPROVED** [review]

File: editor/surface_upgrade_tool.cpp [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+		if (res.is_valid()) {
+			// Ignore things that fail to load.
+			ResourceSaver::save(res);
+		}
+	}
+}
+
+void SurfaceUpgradeTool::_show_popup() {
+	RS::get_singleton()->set_surface_upgrade_callback(nullptr);
+	bool accepted = EditorNode::immediate_confirmation_dialog(TTR("Attempting to use a mesh with an outdated mesh format. The engine needs to update the format in order to use the mesh.\n\nPress 'Upgrade & Re-save' to have the engine scan the project folder and automatically update and re-save all meshes and scenes. This update may take a few minutes. Upgrading will make the meshes incompatible with previous versions of Godot.\n\nPress 'Upgrade Only' to continue opening the scene as normal. The engine will update each mesh in memory, but the update will not be saved. Choosing this option will lead to slower load times every time this project is loaded."), TTR("Upgrade & Re-save"), TTR("Upgrade Only"), 500);
```
Comment (akien-mga) [inline editor/surface_upgrade_tool.cpp:79]: Last nitpick: The first sentence still reads potentially confusing for users, and unnecessarily vague ("a mesh").
Since it's just the trigger, but what's then described and fixed is a project-wide issue, I'd rephrase like that:

```suggestion
	bool accepted = EditorNode::immediate_confirmation_dialog(TTR("This project uses meshes with an outdated mesh format from previous Godot versions. The engine needs to update the format in order to use those meshes.\n\nPress 'Upgrade & Re-save' to have the engine scan the project folder and automatically update and re-save all meshes and scenes. This update may take a few minutes. Upgrading will make the meshes incompatible with previous versions of Godot.\n\nPress 'Upgrade Only' to continue opening the scene as normal. The engine will update each mesh in memory, but the update will not be saved. Choosing this option will lead to slower load times every time this project is loaded."), TTR("Upgrade & Re-save"), TTR("Upgrade Only"), 500);
```

**Comment by akien-mga** [comment]
Tested on https://github.com/team-godog/NGJ2023_Godog and another local prototype, both options work great!

**Comment by fire** [comment]
I approve of the idea and it's necessary to avoid usability problems upgrading the mesh format.

File: editor/surface_upgrade_tool.cpp (hunk 7 above)
Comment (clayjohn) [reply editor/surface_upgrade_tool.cpp]: Updated!

**Comment by akien-mga** [comment]
Thanks!

**Comment by TokisanGames** [comment]
Is there now versioning inside of ArrayMesh?
What Godot versions are meshes upgraded from and to? Only from Godot 4.0 alpha 17 through 4.1.2 upgraded to 4.2?

**Comment by clayjohn** [comment]
> Is there now versioning inside of ArrayMesh?

Yes.

> What Godot versions are meshes upgraded from and to? Only from Godot 4.0 alpha 15 through 4.1.2 upgraded to 4.2?

That's right

**Comment by Lippanon** [comment]
Upgrading engine from v4.2.dev.custom_build [f7bc653cb] to v4.2.beta.custom_build [914445748]
It seems when upgrading a project that has an addon active (GodotJolt in my case), the first time opening the project there will be a prompt about restarting due to addon:
image
On selecting Restart, when the project reopens there will be no upgrade prompt, it automatically chooses to upgrade it seems (I can only tell from the console). That probably shouldn't happen.
On selecting Continue, it will eventually bring the mesh upgrade tool.

I'm upgrading a reasonable size project and it corrupts many scenes, removing everything from the tscn file except the first line. It is corrupting scenes that have no meshes in them but I'll try to gather more information before opening an issue.

**Comment by TokisanGames** [comment]
FYI, only certain GDExtensions like Godot-Jolt and Terrain3D use this prompt. 

What creates the restart prompt is `set_minimum_library_initialization_level(MODULE_INITIALIZATION_LEVEL_SERVERS)` instead of `MODULE_INITIALIZATION_LEVEL_SCENE` in register_types.cpp. Without this prompt, Godot crashes on the first startup, though that may be resolved in 4.2. See #80850 and #81478.

https://github.com/TokisanGames/Terrain3D/pull/196/commits/dc007d4fce425f0f506dca7af0efc60a4d3df493#diff-c23d1074e35b344ac28577f61ecb59e92261b451b50fc76b18eec920384c91b0R40

**Comment by clayjohn** [comment]
@TokisanGames this tool doesn't interact with add-ons at all. If you are having problems with the add-on prompt, I suggest you open an issue about it. 

There are known problems with the surface upgrade tool (see https://github.com/godotengine/godot/issues/83991) and I am fixing them as quickly as I can.
